File: framework/languages/java/e2e/SpotService/Shared/src/main/java/systems/zlink/e2e/spotservice/shared/TimerOverrunHandler.java
package systems.zlink.e2e.spotservice.shared;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

import systems.zlink.framework.spots.ZLinkSpotTimerHandler;
import systems.zlink.framework.spots.ZLinkTimerTick;

public final class TimerOverrunHandler implements ZLinkSpotTimerHandler<TimerScenarioSpot> {
    @Override
    public CompletionStage<Void> handle(
        TimerScenarioSpot spot,
        ZLinkTimerTick tick) {
        try {
            Thread.sleep(160);
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("timer overrun interrupted", error);
        }
        spot.overrunTick(tick.deliveryIndex(), tick.skippedTicks());
        return CompletableFuture.completedFuture(null);
    }
}

File: framework/languages/java/e2e-kotlin/SpotService/Server/Play/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/play/handlers/TimerOverrunHandler.kt
package systems.zlink.e2e.kotlin.spotservice.play.handlers

import systems.zlink.e2e.kotlin.spotservice.Contracts
import systems.zlink.e2e.kotlin.spotservice.ScenarioState
import systems.zlink.e2e.kotlin.spotservice.play.spots.*
import kotlinx.coroutines.delay
import systems.zlink.framework.kotlin.ZLinkSuspendingSpotTimerHandler
import systems.zlink.framework.spots.ZLinkTimerTick

class TimerOverrunHandler : ZLinkSuspendingSpotTimerHandler<TimerScenarioSpot> {
    override suspend fun handle(spot: TimerScenarioSpot, tick: ZLinkTimerTick) {
        delay(160)
        spot.overrunTick(tick.deliveryIndex(), tick.skippedTicks())
    }
}
